Argo-server panics unexpectly while logging.

If you are trying to resolve an environment-specific issue or have a one-off question about the edge case that does not require a feature then please consider asking a question in argocd slack [channel](https://argoproj.github.io/community/join-slack).

Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

See logs

**To Reproduce**

Need to figure out more to reproduce.

**Expected behavior**

No panic

**Screenshots**

**Version**

```shell
➜  ~ argocd version
argocd: v2.0.5+4c94d88
  BuildDate: 2021-07-22T20:54:35Z
  GitCommit: 4c94d886f56bcb2f9d5b3251fdc049c2d1354b88
  GitTreeState: clean
  GoVersion: go1.16
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v2.0.5+4c94d88
```

**Logs**

```go
panic: send on closed channel

goroutine 868 [running]:
github.com/argoproj/argo-cd/v2/server/application.mergeLogStreams.func2(0xc03efa48e7023b01, 0x677ed0795)
	/go/src/github.com/argoproj/argo-cd/server/application/logs.go:114 +0x4d4
github.com/argoproj/argo-cd/v2/server/application.mergeLogStreams.func3(0xc002b30820, 0xc004203218, 0xc002b6b4a0, 0x5f5e100, 0xc002cba060)
	/go/src/github.com/argoproj/argo-cd/server/application/logs.go:128 +0x179
created by github.com/argoproj/argo-cd/v2/server/application.mergeLogStreams
	/go/src/github.com/argoproj/argo-cd/server/application/logs.go:123 +0x2a5
```

Just observed this in 2.13.1, I don't think the above fixed it.

```
panic: send on closed channel

goroutine 446 [running]:
github.com/argoproj/argo-cd/v2/server/application.mergeLogStreams.func2(0x1)
        /go/src/github.com/argoproj/argo-cd/server/application/logs.go:114 +0x46c
github.com/argoproj/argo-cd/v2/server/application.mergeLogStreams.func3()
        /go/src/github.com/argoproj/argo-cd/server/application/logs.go:128 +0x113
created by github.com/argoproj/argo-cd/v2/server/application.mergeLogStreams in goroutine 368
        /go/src/github.com/argoproj/argo-cd/server/application/logs.go:123 +0x2eb
```

--- a/server/application/logs.go
+++ b/server/application/logs.go
@@ -120,16 +120,22 @@ func mergeLogStreams(streams []chan logEntry, bufferingDuration time.Duration) c
 	var sentAt time.Time
 
 	ticker := time.NewTicker(bufferingDuration)
+	done := make(chan struct{})
 	go func() {
-		for range ticker.C {
-			sentAtLock.Lock()
-			// waited long enough for logs from each streams, send everything accumulated
-			if sentAt.Add(bufferingDuration).Before(time.Now()) {
-				_ = send(true)
-				sentAt = time.Now()
-			}
+		for {
+			select {
+			case <-done:
+				return
+			case <-ticker.C:
+				sentAtLock.Lock()
+				// waited long enough for logs from each streams, send everything accumulated
+				if sentAt.Add(bufferingDuration).Before(time.Now()) {
+					_ = send(true)
+					sentAt = time.Now()
+				}
 
-			sentAtLock.Unlock()
+				sentAtLock.Unlock()
+			}
 		}
 	}()
 
@@ -145,6 +151,11 @@ func mergeLogStreams(streams []chan logEntry, bufferingDuration time.Duration) c
 		_ = send(true)
 
 		ticker.Stop()
+		// ticker.Stop() does not close the channel, and it does not wait for the channel to be drained. So we need to
+		// explicitly prevent the gorountine from leaking by closing the channel. We also need to prevent the goroutine
+		// from calling `send` again, because `send` pushes to the `merged` channel which we're about to close.
+		// This describes the approach nicely: https://stackoverflow.com/questions/17797754/ticker-stop-behaviour-in-golang
+		done <- struct{}{}
 		close(merged)
 	}()
 	return merged

--- a/server/application/logs_test.go
+++ b/server/application/logs_test.go
@@ -96,9 +96,12 @@ func TestMergeLogStreams_RaceCondition(t *testing.T) {
 
 		merged := mergeLogStreams([]chan logEntry{first, second}, 1*time.Millisecond)
 
-		var lines []string
-		for entry := range merged {
-			lines = append(lines, entry.line)
+		// Drain the channel
+		for range merged {
 		}
+
+		// This test intentionally doesn't test the order of the output. Under these intense conditions, the test would
+		// fail often due to out of order entries. This test is only meant to reproduce a race between a channel writer
+		// and channel closer.
 	}
 }

--- a/server/application/logs_test.go
+++ b/server/application/logs_test.go
@@ -78,23 +78,23 @@ func TestMergeLogStreams(t *testing.T) {
 
 func TestMergeLogStreams_RaceCondition(t *testing.T) {
 	// Test for regression of this issue: https://github.com/argoproj/argo-cd/issues/7006
-	for i := 0; i < 100; i++ {
+	for i := 0; i < 1000; i++ {
 		first := make(chan logEntry)
 		second := make(chan logEntry)
 
 		go func() {
 			parseLogsStream("first", io.NopCloser(strings.NewReader(`2021-02-09T00:00:01Z 1`)), first)
-			time.Sleep(time.Duration(i%5) * time.Millisecond)
+			time.Sleep(time.Duration(i%10) * time.Millisecond)
 			close(first)
 		}()
 
 		go func() {
 			parseLogsStream("second", io.NopCloser(strings.NewReader(`2021-02-09T00:00:02Z 2`)), second)
-			time.Sleep(time.Duration((i+2)%5) * time.Millisecond)
+			time.Sleep(time.Duration((i+5)%10) * time.Millisecond)
 			close(second)
 		}()
 
-		merged := mergeLogStreams([]chan logEntry{first, second}, 1*time.Millisecond)
+		merged := mergeLogStreams([]chan logEntry{first, second}, 5*time.Millisecond)
 
 		var lines []string
 		for entry := range merged {

--- a/server/application/logs_test.go
+++ b/server/application/logs_test.go
@@ -78,30 +78,27 @@ func TestMergeLogStreams(t *testing.T) {
 
 func TestMergeLogStreams_RaceCondition(t *testing.T) {
 	// Test for regression of this issue: https://github.com/argoproj/argo-cd/issues/7006
-	for i := 0; i < 1000; i++ {
+	for i := 0; i < 5000; i++ {
 		first := make(chan logEntry)
 		second := make(chan logEntry)
 
 		go func() {
 			parseLogsStream("first", io.NopCloser(strings.NewReader(`2021-02-09T00:00:01Z 1`)), first)
-			time.Sleep(time.Duration(i%10) * time.Millisecond)
+			time.Sleep(time.Duration(i%3) * time.Millisecond) // Reduce the variability in timing
 			close(first)
 		}()
 
 		go func() {
 			parseLogsStream("second", io.NopCloser(strings.NewReader(`2021-02-09T00:00:02Z 2`)), second)
-			time.Sleep(time.Duration((i+5)%10) * time.Millisecond)
+			time.Sleep(time.Duration((i+1)%3) * time.Millisecond)
 			close(second)
 		}()
 
-		merged := mergeLogStreams([]chan logEntry{first, second}, 5*time.Millisecond)
+		merged := mergeLogStreams([]chan logEntry{first, second}, 1*time.Millisecond)
 
 		var lines []string
 		for entry := range merged {
 			lines = append(lines, entry.line)
 		}
-
-		expected := []string{"1", "2"}
-		assert.Equal(t, expected, lines)
 	}
 }

--- a/server/application/logs_test.go
+++ b/server/application/logs_test.go
@@ -75,3 +75,33 @@ func TestMergeLogStreams(t *testing.T) {
 
 	assert.Equal(t, []string{"1", "2", "3", "4"}, lines)
 }
+
+func TestMergeLogStreams_RaceCondition(t *testing.T) {
+	// Test for regression of this issue: https://github.com/argoproj/argo-cd/issues/7006
+	for i := 0; i < 100; i++ {
+		first := make(chan logEntry)
+		second := make(chan logEntry)
+
+		go func() {
+			parseLogsStream("first", io.NopCloser(strings.NewReader(`2021-02-09T00:00:01Z 1`)), first)
+			time.Sleep(time.Duration(i%5) * time.Millisecond)
+			close(first)
+		}()
+
+		go func() {
+			parseLogsStream("second", io.NopCloser(strings.NewReader(`2021-02-09T00:00:02Z 2`)), second)
+			time.Sleep(time.Duration((i+2)%5) * time.Millisecond)
+			close(second)
+		}()
+
+		merged := mergeLogStreams([]chan logEntry{first, second}, 1*time.Millisecond)
+
+		var lines []string
+		for entry := range merged {
+			lines = append(lines, entry.line)
+		}
+
+		expected := []string{"1", "2"}
+		assert.Equal(t, expected, lines)
+	}
+}

--- a/server/application/logs_test.go
+++ b/server/application/logs_test.go
@@ -84,7 +84,7 @@ func TestMergeLogStreams_RaceCondition(t *testing.T) {
 
 		go func() {
 			parseLogsStream("first", io.NopCloser(strings.NewReader(`2021-02-09T00:00:01Z 1`)), first)
-			time.Sleep(time.Duration(i%3) * time.Millisecond) // Reduce the variability in timing
+			time.Sleep(time.Duration(i%3) * time.Millisecond)
 			close(first)
 		}()
 

